lib/matplotlib/tests/test_image.py

```diff
@@ -1083,3 +1083,17 @@ def test_deprecation():
         with pytest.warns(MatplotlibDeprecationWarning):
             # Enough arguments to pass "shape" positionally.
             obj.imshow(data, *[None] * 10)
+
+
+@image_comparison(baseline_images=['image_array_alpha'], remove_text=True)
```

Comment:
Consider using check_figures_equal (https://matplotlib.org/devdocs/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal) instead: not only this avoids adding baseline images to the repo, but this also actually checks the desired behavior, which is that per-pixel alpha should be equivalent to passing a RGBA array with the correct alpha to start with.

Reply:
Good point, will tackle that after we decide on whether to go with this implementation or the variant in https://github.com/tillahoffmann/matplotlib/pull/1.

Reply:
Feel free to force-push anything in a PR (my PRs, unless huge, are mostly always a single commit that's always force-pushed over after reviews) -- especially to clear out the reference images :)

Reply:
I gave the `check_figures_equal` test approach a go but didn't quite manage to get it to work: the interpolation is working out ever so slightly differently (see below). Possible next steps:

* Stick with `@image_comparison`
* Increase tolerance
* Try to get the interpolation to work the same

Suggestions on how to proceed?

```python
@check_figures_equal()
def test_image_array_alpha(fig_test, fig_ref):
    '''per-pixel alpha channel test'''
    x = np.linspace(0, 1)
    xx, yy = np.meshgrid(x, x)

    zz = np.exp(- 3 * ((xx - 0.5) ** 2) + (yy - 0.7 ** 2))
    alpha = zz / zz.max()

    cmap = plt.get_cmap('viridis')
    ax = fig_test.add_subplot(111)
    ax.imshow(zz, alpha=alpha, cmap=cmap)

    ax = fig_ref.add_subplot(111)
    rgba = cmap(colors.Normalize()(zz))
    rgba[..., -1] = alpha
    ax.imshow(rgba)
```

## Actual
![test_image_array_alpha](https://user-images.githubusercontent.com/966348/63495103-1ca2c780-c4b7-11e9-81c3-f88f79f1b4d4.png)

## Expected
![test_image_array_alpha-expected](https://user-images.githubusercontent.com/966348/63495122-23c9d580-c4b7-11e9-9e05-ffd5fa7fcc36.png)

## Difference
![test_image_array_alpha-failed-diff](https://user-images.githubusercontent.com/966348/63495151-317f5b00-c4b7-11e9-8ec0-b066d7f18e83.png)

Reply:
As annoying as this may sound I think we should figure out why the two approaches give different results before merging this pr...

Reply:
What *is* the difference?  The difference plot just looks black to me...

Reply:
you have to open the file and examine the contents (e.g. with np.nonzero) to find some minute differences.

Reply:
Is the normalization the same?  You call `colors.Normalize()`, maybe that's a bit different?

Reply:
I've dug into this a bit more and the unexpected behaviour seems to be unrelated to the code changes in `image.py` (but maybe the test is badly written). In particular, I checked out the current `master` (9a1873f0462af8b45b2fe16c37209c9e0af78b7f) and added the following test to `test_image.py`, which fails when run using `py.test lib/matplotlib/tests/test_image.py -k test_image_array_alpha -x --pdb` because there are small differences between the images. Note that I have not made any other changes and am not using any alpha channels.

```python
@check_figures_equal()
def test_image_array_alpha(fig_test, fig_ref):
    x = np.linspace(0, 1)
    xx, yy = np.meshgrid(x, x)
    zz = np.exp(- 3 * ((xx - 0.5) ** 2) + (yy - 0.7 ** 2))
    alpha = zz / zz.max()
    cmap = plt.get_cmap()
    norm = colors.Normalize()

    ax1 = fig_test.add_subplot(111)
    ax1.imshow(zz, cmap=cmap)

    ax2 = fig_ref.add_subplot(111)
    rgba = cmap(norm(zz))
    ax2.imshow(rgba)
```

I added the following snippet at the end of `_make_image` to check the content of the RGBA image produced.

```python
import hashlib
for i in range(output.shape[-1]):
    print(i, hashlib.sha1(np.ascontiguousarray(output[..., i])).hexdigest())
```

As might be expected, the hashes of the individual channels differ.

```
0 bd8859ed65700863f9b8b46322d3845bd95b2f07
1 691db54c0d3e3b086bb5c95eff650fa8c6b55c55
2 b7718c7b88e40b67f2cdf7c6e6935aa38e06f2e2
3 b2fa672f8bba0b9c504e83c5b17ac848f0c14977

0 8114943151ea1dc85ae4f9db5c0f6bfcdb2dd50d
1 d791e226af46b1d7f61b27974168f656b1d1924c
2 7918620835c6b94a4e5325670cd87ac39969180c
3 b2fa672f8bba0b9c504e83c5b17ac848f0c14977
```

Another odd behaviour: if I run the code as a stand-alone script the hashes for the different channels are identical. In particular, I added the following to `test_image.py` and temporarily removed the `@check_figures_equal()` decorator.

```
if __name__ == '__main__':
    fig_test = plt.figure()
    fig_ref = plt.figure()
    test_image_array_alpha(fig_test, fig_ref)

    import io
    fig_test.savefig(io.BytesIO())
    fig_ref.savefig(io.BytesIO())
```

Executing with `python lib/matplotlib/tests/test_image.py` gives the following output.

```
0 8558f3e6bff868721d60f2e2f667dabb8a55ca7f
1 1c73506069b49d25c31cfd826e3fe6e0788ad92a
2 68c38d4a754878d6aaa31b0a21d12e83f58b8812
3 c69dcd9b189359174ecedd51bc7ca1c42a430605

0 8558f3e6bff868721d60f2e2f667dabb8a55ca7f
1 1c73506069b49d25c31cfd826e3fe6e0788ad92a
2 68c38d4a754878d6aaa31b0a21d12e83f58b8812
3 c69dcd9b189359174ecedd51bc7ca1c42a430605
```

Any ideas?

Reply:
Perhaps set `interpolation="none"` or `"nearest"`?  The default *in tests* is to use classic-style, which does perform interpolation, but here it may be the source of the problem.

Reply:
Yes, the tests pass on `master` if I set `interpolation='nearest'` for both `imshow` calls. I'll try to get the same to run with the alpha channels enabled.